### PR TITLE
Introduce MM communicate and variable service into MU_TIANO_ARM

### PR DESCRIPTION
**Pull request by kuqin12**
This change introduces 2 new modules into MU_TIANO_ARM repo, to support the MM communication and variable read service in PEI phase on ARM based platforms.

### CODE REVIEW

**Review by os-d — APPROVED** [review]

**Comment by makubacki** [comment]
Since it's end of day, please give me some time tomorrow to come back to this.

File: ArmPkg/Drivers/SmmVariablePei/SmmVariablePei.c [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+  @retval EFI_DEVICE_ERROR      The variable could not be retrieved because of a device error.
+
+**/
+EFI_STATUS
+EFIAPI
+PeiSmmGetVariable (
+  IN CONST  EFI_PEI_READ_ONLY_VARIABLE2_PPI  *This,
+  IN CONST  CHAR16                           *VariableName,
+  IN CONST  EFI_GUID                         *VariableGuid,
+  OUT       UINT32                           *Attributes,
```
Comment (makubacki) [inline ArmPkg/Drivers/SmmVariablePei/SmmVariablePei.c]:
```suggestion
  OUT       UINT32                           *Attributes,  OPTIONAL
```

File: ArmPkg/Drivers/SmmVariablePei/SmmVariablePei.c [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+  @param  VariableGuid          A pointer to an EFI_GUID that is the variable's GUID. The combination of
+                                VariableGuid and VariableName must be unique.
+  @param  Attributes            If non-NULL, on return, points to the variable's attributes.
+  @param  DataSize              On entry, points to the size in bytes of the Data buffer.
+                                On return, points to the size of the data returned in Data.
+  @param  Data                  Points to the buffer which will hold the returned variable value.
+                                May be NULL with a zero DataSize in order to determine the size of the buffer needed.
+
+  @retval EFI_SUCCESS           The variable was read successfully.
+  @retval EFI_NOT_FOUND         The variable was be found.
```
Comment (makubacki) [inline ArmPkg/Drivers/SmmVariablePei/SmmVariablePei.c]:
```suggestion
  @retval EFI_NOT_FOUND         The variable was not found.

```

File: ArmPkg/Drivers/SmmVariablePei/SmmVariablePei.c [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+  SMM_VARIABLE_COMMUNICATE_ACCESS_VARIABLE  *SmmVarAccessHeader;
+  UINT8                                     *MmCommunicateBuffer;
+  UINT8                                     RequiredPages;
+  EFI_PEI_MM_COMMUNICATION_PPI              *MmCommunicationPpi;
+
+  // Check input parameters
+  if ((VariableName == NULL) || (VariableGuid == NULL) || (DataSize == NULL)) {
+    return EFI_INVALID_PARAMETER;
+  }
+
```
Comment (makubacki) [inline ArmPkg/Drivers/SmmVariablePei/SmmVariablePei.c:97]: Worth considering a

```c
  if (VariableName[0] == 0) {
    return EFI_NOT_FOUND;
  }
```

To match the normal implementation.

File: ArmPkg/Drivers/SmmVariablePei/SmmVariablePei.c [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+    DEBUG ((DEBUG_ERROR, "%a: Failed to locate PEI MM Communication PPI: %r\n", __FUNCTION__, Status));
+    return Status;
+  }
+
+  // Allocate required pages to send MM request
+  RequiredPages = ((sizeof (EFI_MM_COMMUNICATE_HEADER) + \
+                    sizeof (SMM_VARIABLE_COMMUNICATE_HEADER) + \
+                    sizeof (SMM_VARIABLE_COMMUNICATE_ACCESS_VARIABLE) + \
+                    StrSize (VariableName) + *DataSize) \
+                   /EFI_PAGE_SIZE) + 1;
```
Comment (makubacki) [inline ArmPkg/Drivers/SmmVariablePei/SmmVariablePei.c]: Any reason not to use `EFI_SIZE_TO_PAGES()` here?

File: ArmPkg/Drivers/MmCommunicationPei/MmCommunicationPei.inf [hunk 5]
```diff
@@ -0,0 +1,40 @@
+#/** @file -- MmCommunicationPei.inf
+#
+#  Copyright (c) 2011-2014, ARM Ltd. All rights reserved.<BR>
```
Comment (makubacki) [inline ArmPkg/Drivers/MmCommunicationPei/MmCommunicationPei.inf]: Is this really straight from some old Arm code?

Comment (makubacki) [reply ArmPkg/Drivers/MmCommunicationPei/MmCommunicationPei.inf]: The other files in the module have "Copyright (c) Microsoft Corporation. All rights reserved." so wanted to double check.

**Review by makubacki — APPROVED** [review]
Left some comments but overall looks fine

File: ArmPkg/Drivers/SmmVariablePei/SmmVariablePei.c (hunk 3 above)
Comment (kuqin12) [reply ArmPkg/Drivers/SmmVariablePei/SmmVariablePei.c:97]: Thanks, this is added.

File: ArmPkg/Drivers/SmmVariablePei/SmmVariablePei.c (hunk 1 above)
Comment (kuqin12) [reply ArmPkg/Drivers/SmmVariablePei/SmmVariablePei.c]: Updated.

File: ArmPkg/Drivers/SmmVariablePei/SmmVariablePei.c (hunk 2 above)
Comment (kuqin12) [reply ArmPkg/Drivers/SmmVariablePei/SmmVariablePei.c]: Updated

File: ArmPkg/Drivers/SmmVariablePei/SmmVariablePei.c (hunk 4 above)
Comment (kuqin12) [reply ArmPkg/Drivers/SmmVariablePei/SmmVariablePei.c]: Agreed. This is updated.

File: ArmPkg/Drivers/MmCommunicationPei/MmCommunicationPei.inf (hunk 5 above)
Comment (kuqin12) [reply ArmPkg/Drivers/MmCommunicationPei/MmCommunicationPei.inf]: Thanks. The copyrights are fixed.

Comment (kuqin12) [reply ArmPkg/Drivers/MmCommunicationPei/MmCommunicationPei.inf]: The MSFT copyrights and the original ones are both updated/added.

**Comment by kuqin12** [comment]
I think some oddness happened with the PR UI. The commits are merged as "Rebased and FF". Closing this manually to avoid confusion.

**Pull request closed without merging**